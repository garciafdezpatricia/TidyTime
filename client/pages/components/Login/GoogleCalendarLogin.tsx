import {useGoogleLogin} from "@react-oauth/google";
import { useEffect, useState } from "react";

interface interfaceToken {
    access_token: string,
    expires_in: number,
    id_token: string,
    refresh_token: string,
    scope: string,
    token_type: string,
}

export default function LoginGoogleCalendar() {
    const [isLoggedIn, setIsLoggedIn] = useState(false);
    const [tokenslocal, setTokensLocal] = useState<interfaceToken>()

    useEffect(() => {
        if (localStorage.getItem("GTokens")){
            setTokensLocal(JSON.parse(localStorage.getItem("GTokens") as string))
        }   
    }, [isLoggedIn])

    const handleLoginSuccess = (tokens:any) => {
        //TODO: securely save them
        localStorage.setItem("GTokens", JSON.stringify(tokens));
        setIsLoggedIn(true);
    }

    const googleLogin = useGoogleLogin({
        onSuccess: (tokenResponse) => {
            fetch('http://localhost:8080/api/auth/google', {
                method: 'POST',
                headers: {
                  'Content-Type': 'application/json',
                },
                body: JSON.stringify({ code: tokenResponse.code }),
              })
              .then(response => response.json())
              .then(data => {
                console.log('Backend response:', data);
                handleLoginSuccess(data);
              })
              .catch(error => {
                console.error('Error:', error);
              });
        },
        onError: () => {
            console.log("Error login");
        },
        flow: 'auth-code',
        scope: 'https://www.googleapis.com/auth/calendar openid https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email',
    })

    return (
        <>
            { !isLoggedIn && 
                <button 
                    onClick={() => googleLogin()}
                >Sign in with Google
                </button>
            }
        </>
    )
}

    // function getCalendar() {
    //     if (tokenslocal) {
    //         fetch('http://localhost:8080/google/calendar', {
    //             method: 'GET',
    //             headers: {
    //                 'Authorization': 'Bearer ' + tokenslocal.access_token
    //             }
    //         })
    //         .then(response => response.json())
    //         .then(data => console.log(data))
    //     }
    // } 