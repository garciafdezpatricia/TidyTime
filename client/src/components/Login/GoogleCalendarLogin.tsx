import { useEffect, useState } from "react";
import { Icon } from "../Icon/Icon";
import { useGoogleContext } from "../Context/GoogleContext";
import { useGoogleHandler } from "@/pages/api/google";
import { useTranslation } from "react-i18next";

export default function LoginGoogleCalendar() {
    const { t } = useTranslation();
    const {loggedIn, authUrl} = useGoogleContext();
    const { handleLogin, handleLogout} = useGoogleHandler();

    const [logginInGoogle, setLogginInGoogle] = useState(false);

    useEffect(() => {
        if (authUrl !== "" && !loggedIn && logginInGoogle) {
            window.location.assign(authUrl);
        }
    }, [authUrl, loggedIn]);

    const onLogout = () => {
        handleLogout();
    }

    const onConnect = () => {
        // there's no need to set it to false afterwards since the app is going to be redirected
        // to another page which will redirect the application to this page (and loggingin will be false)
        setLogginInGoogle(true);
        handleLogin();
    }

    return (
        <> 
        { 
            !loggedIn
            ?
            <button className="google-login-button" onClick={onConnect}>
                <Icon src={"./google.svg"} alt={"Connect to Google"} />
                {t('calendar.connectGoogle')}
            </button>
            :
            <button className="google-logout-button" onClick={() => onLogout()}>
                <Icon src={"./google.svg"} alt={"Log out of Google"} />
                {t('home.logoutButton')}
            </button>
        }
        </>
    )
}