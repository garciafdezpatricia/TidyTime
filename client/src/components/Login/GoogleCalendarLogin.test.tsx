import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import GoogleCalendarLogin from '@/src/components/Login/GoogleCalendarLogin';

describe('GoogleCalendarLogin', () => {
    it('renders without crashing', () => {
      const { container } = render(<GoogleCalendarLogin/>);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<GoogleCalendarLogin/>);
        expect(asFragment()).toMatchSnapshot();
      });
  });