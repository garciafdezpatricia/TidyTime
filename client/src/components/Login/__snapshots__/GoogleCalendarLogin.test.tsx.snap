// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleCalendarLogin matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="google-login-button"
  >
    <img
      alt="Connect to Google"
      class="btn-icon"
      src="./google.svg"
    />
    calendar.connectGoogle
  </button>
</DocumentFragment>
`;
